taskqueue/taskqueue.go

```diff
@@ -93,13 +93,11 @@ func (tq *WorkerTaskQueue) Shutdown() {
 }
 
 func (tq *WorkerTaskQueue) WaitForNoActiveTasks() {
-	for atomic.LoadInt32(&tq.activeTasks) > 0 {
-		select {
-		case <-tq.ctx.Done():
-			return
-		case <-tq.noTaskSignal:
-		}
+	tq.noTaskCond.L.Lock()
+	for tq.activeTasks > 0 {
```

Comment:
I know this looks unnecessary, but my assumption here is that there are no guarantees as to who gets the next mutex lock if there are multiple parties waiting on it - both via `Wait()` and `Lock()`. So if a second task started and got the lock before our wait caller got a mutex notify then we'd not be in a `tq.activeTasks>0` condition even after waking up here.

At least that's how it would work in C++ and Java. Or do we have stronger guarantees in Go about who gets the lock next after a `Broadcast()`?

Reply:
I think the loop is right - the sync.Cond.Wait docs use a loop as an example too.